### PR TITLE
gc: add clear range for full range span when it is covered by range tombstone

### PR DESCRIPTION
**Pull request by aliher1911**
This commit adds range_deletion_keys parameters to GC request.
range_deletion_keys is used by GC queue to remove chunks of
consecutive keys where no new data was written above the GC
threshold and storage can optimize deletions with range
tombstones.
To support new types of keys, GCer interface is also updated
to pass provided keys down to request.

---

storage: add gc operation that uses clear range
GC relies on point and range key deletion operations
when removing data. This is not good in cases where full
range is deleted e.g. schema change.
This commit add a clear operation that would remove all
range data if it is completely covered by range tombstones. And
range tombstones themselves are below gc threshold.

---

rditer: skip option for range user key space
Option to skip user key space to allow fast path GC
to avoid unnecessary scans.

---

gc: add fast path check to identify empty ranges
This commit adds a fast path check for ranges where
all data is covered by a range tombstone below gc ttl
threshold. If such range is identified, GC will perform
GC Clear Range of the full range to avoid rescanning
range and recomputing stats.

---

batcheval: add clear range option to cmd_gc
GC request will perform fast path clear operation on the
whole range if instructed by GC. Its range bounds should
match descriptor bounds exactly for operation to succeed.
Underlying mvcc operation will validate if no data is
present for safety.

### CODE REVIEW

File: pkg/kv/kvserver/gc/gc_random_test.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/gc/gc_random_test.go:736]: nit: `gcClearRangeKeys`

File: pkg/storage/mvcc_test.go
Comment (erikgrinaker) [inline pkg/storage/mvcc_test.go:5997]: I'd like some tests of inline values too, in particular that we won't accidentally GC them even though they're technically below the GC threshold (timestamp 0).

File: pkg/storage/mvcc.go
Comment (erikgrinaker) [inline pkg/storage/mvcc.go:5197]: I wonder if we should exclude system data from this fast path too, just to be safe. I.e. anything below `SystemMax`.

File: pkg/kv/kvserver/gc/gc.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/gc/gc.go:384]: If we're returning here, won't we bypass GC of local data?

File: pkg/kv/kvserver/gc/gc.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/gc/gc.go:381]: Should we log an error if this fails?

File: pkg/kv/kvserver/gc/gc.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/gc/gc.go:320]: This is sort of true, but only because we don't currently write range keys across local keyspace. The fast path only applies to global keys. Nothing really prevents us from writing MVCC range tombstones across local keyspace. Should we call into this path anyway?

File: pkg/kv/kvserver/metrics.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/metrics.go:1354]: nit: "Number of successful ClearRange operations during GC".

Let's track failures too.

**Review by erikgrinaker — COMMENTED** [review]
This fast path often won't be used in serverless, because tables don't fall along range boundaries. Large tables that span multiple ranges will cover entire ranges though. So in the worst case we'll have two ranges with close to 512 MB that needs to be cleared key-by-key. That's unfortunate, but doesn't seem _terrible_. Wdyt @ajwerner?

File: pkg/storage/mvcc.go
Comment (aliher1911) [reply pkg/storage/mvcc.go:5175]: From my understanding for rditer there are range local spans and user data spans. All meta, liveness etc are "user data" as in global data accessible by kv requests vs. data that is updated implicitly as part of kv operations. So we currently use range tombstones on "user data" but never on lock tables and the like. And this is also a divide of uses vs system data in mvcc stats.

File: pkg/kv/kvserver/gc/gc.go
Comment (aliher1911) [reply pkg/kv/kvserver/gc/gc.go:320]: We can just pass down if we want to exclude spans that we covered by clear range. Currently it is only to userspan but if we extend range keys and clear range to others we might need to split iteration over keys which was done in unfinished batch pr 😢

File: pkg/storage/mvcc.go
Comment (erikgrinaker) [reply pkg/storage/mvcc.go:5175]: Yup. 👍

File: pkg/kv/kvserver/gc/gc.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/gc/gc.go:320]: Passing down user spans seems fine.

File: pkg/storage/mvcc_test.go
Comment (aliher1911) [reply pkg/storage/mvcc_test.go:5998]: Aren't those values covered by intents? They should have 0 timestamp. At least that's how GC processes them on the queue side. I'll add specific test just in case.

File: pkg/storage/mvcc.go
Comment (aliher1911) [reply pkg/storage/mvcc.go]: Thanks for  pointing this. I started with incremental and time bounds, but masking option seem more viable here or at least simpler.

File: pkg/kv/kvserver/gc/gc.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/gc/gc.go]: I assume so, otherwise I don't understand why we bother with iterating over them here.

Comment (erikgrinaker) [reply pkg/kv/kvserver/gc/gc.go]: Range descriptors, for example, most certainly use transactions and thus MVCC, and will need GC.

File: pkg/storage/mvcc_test.go
Comment (erikgrinaker) [reply pkg/storage/mvcc_test.go:6002]: Sort of, but the access paths are different enough that it's worth covering both. Intents are stored in a separate lock table, and either require use of an interleaving iterator or by targeting the lock table explicitly, while inline values can show up anywhere.

**Comment by erikgrinaker** [comment]
Looks like we have a bunch of test failures -- can you look at those before I review again?

**Comment by aliher1911** [comment]
I just rebased as some of them seemed to be panics in unrelated parts. Lets see how it flies.

File: pkg/kv/kvserver/batcheval/cmd_gc.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/batcheval/cmd_gc.go:188]: This is going to happen with range merges. The GCer will fall back to point GC, right? We may want to return a structured error and retry it, although I suppose we'd have to bump it all the way back up to the GC queue and wait for it to run the merged range, in which case we could run the risk of never GCing the range if we, for some reason, ended up in a retry loop.

This doesn't seem very likely to happen, since most ranges will have cooled down from e.g. load-based splits by the time the GC TTL hits, so I don't think it's a huge deal.

File: pkg/kv/kvserver/mvcc_gc_queue.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/mvcc_gc_queue.go]: Oof, yeah, passing all of these empty structs everywhere is pretty annoying. I think I was wrong here, sorry about that.

File: pkg/kv/kvserver/metrics.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/metrics.go:1353]: Maybe `clearrangesuccess`, like with `resolve`?

File: pkg/kv/kvserver/rditer/replica_data_iter.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/rditer/replica_data_iter.go:30]: I don't think the bit about fast path/ClearRange belongs here, this iterator is agnostic to who is using it.

File: pkg/kv/kvserver/rditer/replica_data_iter.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/rditer/replica_data_iter.go:114]: We're not including the user key span.

File: pkg/roachpb/api.proto
Comment (erikgrinaker) [inline pkg/roachpb/api.proto:1013]: Let's call out that this needs to cover a complete Raft range.

File: pkg/storage/mvcc.go
Comment (erikgrinaker) [inline pkg/storage/mvcc.go:5175]: Let's add in the intent fields too. Similarly to e.g. `LiveCount`, we expect these to always be 0, but in case the stats are estimates (and thus wrong) we may as well empty them out. We should mention that in a comment.

File: pkg/storage/mvcc_test.go
Comment (erikgrinaker) [inline pkg/storage/mvcc_test.go:5891]: Maybe we should just move all of these over to `TestMVCCHistories`, which seems more convenient? Fine with deferring that until stability too.

File: pkg/kv/kvserver/batcheval/cmd_gc.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/batcheval/cmd_gc.go:186]: We should also check that we're not setting `Keys` or `RangeKeys` when we're processing this.

And at the start of this function, where we're preventing setting the GC threshold when we perform GC operations, we need to guard against `ClearRangeKey` too.

**Review by erikgrinaker — APPROVED** [review]
LGTM once comments are addressed.

**Comment by erikgrinaker** [comment]
By the way, we need to make sure this is appropriately version gated. Is there a chance we can send e.g. a `ClearRangeKey` request if we encounter an empty range or something?

File: pkg/storage/mvcc_test.go
Comment (aliher1911) [reply pkg/storage/mvcc_test.go:5894]: They are all covered by mvcc_test_histories, I removed redundant ones. I'll keep the test with a single basic case, it is much easier to repro or check why certain things didn't work compared to doing it with datadriven.

**Comment by aliher1911** [comment]
> By the way, we need to make sure this is appropriately version gated. Is there a chance we can send e.g. a ClearRangeKey request if we encounter an empty range or something?

I added an explicit check for range tombstones covering data and fail fast path check if none were found. I think that should be enough to gate the empty range case. Thanks for pointing this out!

**Pull request merged**
